extensions/amp-video/1.0/video-wrapper.js

```diff
@@ -0,0 +1,210 @@
+      playerRef.current.pause();
+    });
+  };
+
+  useEffect(() => {
+    if (!mediasession || !metadata || !playing) {
+      return;
+    }
+    setMediaSession(window, metadata, play, pause);
+  }, [mediasession, playing, metadata]);
+
+  return (
+    <div
+      ref={wrapperRef}
+      style={{position: 'relative', width: '100%', height: '100%'}}
+    >
+      <Component
+        ref={playerRef}
+        // Ensure that props controlled here are excluded from passthroughProps,
+        // see destructuring statement above.
+        muted={muted}
+        controls={controls && (!autoplay || userInteracted)}
+        onLoad={() => readyDeferredRef.current.resolve()}
+        onLoadedMetadata={() => {
+          if (!mediasession) {
+            return;
+          }
+
+          const metadata = playerRef.current.getMetadata() || {
+            ...EMPTY_METADATA,
+            artwork: undefined,
+          };
+
+          metadata.title =
+            metadata.title ||
+            passthroughProps.title ||
+            passthroughProps['aria-label'] ||
+            document.title;
+
+          metadata.artwork =
+            metadata.artwork ||
+            parseSchemaImage(document) ||
+            parseOgImage(document) ||
+            parseFavicon(document);
+
+          setMetadata(metadata);
+        }}
+        onPlaying={() => setPlaying(true)}
+        onPause={() => setPlaying(false)}
+        {...passthroughProps}
+      >
+        {children}
+      </Component>
+      {autoplay && !userInteracted && (
+        <Autoplay
+          playing={playing}
+          displayIcon={!noaudio && muted}
+          wrapperRef={wrapperRef}
+          play={play}
+          pause={pause}
+          displayOverlay={controls}
+          onOverlayClick={() => {
+            setMuted(false);
+            setUserInteracted(true);
+          }}
+        />
+      )}
+    </div>
+  );
+}
+
+/**
+ * @param {!VideoAutoplayProps} props
+ * @return {PreactDef.Renderable}
+ */
+function Autoplay({
+  displayIcon,
+  playing,
+  displayOverlay,
+  onOverlayClick,
+  wrapperRef,
+  play,
+  pause,
+}) {
+  useEffect(() => {
+    let observer = new IntersectionObserver(
+      (entries) => {
+        if (entries[0].isIntersecting) {
+          play();
+        } else {
+          pause();
+        }
+      },
+      {threshold: MIN_VISIBILITY_RATIO_FOR_AUTOPLAY}
+    );
+
+    observer.observe(wrapperRef.current);
+
+    return () => {
+      observer.disconnect();
+      observer = null;
+    };
+  }, [wrapperRef, play, pause]);
+
+  return (
+    <>
+      {displayIcon && (
+        <div
+          // TODO(alanorozco): animation-play-state won't work on iOS, this
+          // needs an `unpausable` classname to optionally remove animation on
+          // platform.
+          class={`amp-video-eq ${playing ? `amp-video-eq-play` : ''}`}
+          style={{display: 'flex'}}
+        >
+          {[1, 2, 3, 4].map((i) => (
```

Comment:
Why 4?

Reply:
Four animated columns.

![image](https://user-images.githubusercontent.com/254946/88697055-e6070100-d0b8-11ea-8eac-eb51885e87df.png)

Reply:
IMHO this code feels more like an SVG to me. What I mean by this is that it's a mostly external resource to the main code. It'd be good to put it in a constant and just output here. I think that'd be possible, right? Generally, in React, I'd put it in a separate component and then use `React.memo()` to turn it in a sort-of-constant. Not sure if Preact has a similar option.

Reply:
`memo` is part of `preact/compat`. From a quick glance it seems like it can be DCE'd.

There's a few options here:

1. Setting a module constant, either directly or lazy with `once()`.
2. `useMemo` for the constant part of the icon (the columns). (I went for this at the moment.)
3. `memo()`. Though that still re-renders on prop change, e.g. `playing`.

Reply:
I think the module constant with just JSX should work. I don't see any reason right now that it wouldn't. And by extension some sort of `once()` will work as well. So I think either is a good idea. I think `useMemo` would be ok as well. But agree - `memo()` is probably not worth it.